### PR TITLE
Replace multiclass mcc with 1-mcc for loss

### PR DESCRIPTION
**Pull request by cjmcgill**
## Bugfix / Desired workflow
Problem is well-explained in #331. The Matthews Correlation Coefficient for multiclass jobs is described in the sklearn documentation (https://scikit-learn.org/stable/modules/model_evaluation.html#matthews-correlation-coefficient), The MCC should be maximized for a high quality model. However as currently implemented, it is set as `loss = MCC` so it is minimized. The change in this PR is to change it to `loss = 1 - MCC` instead, so the MCC will be maximized. This is already how MCC for classification is set up but not multiclass.

### CODE REVIEW

File: chemprop/train/loss_functions.py [hunk 1]
```diff
@@ -148,7 +148,8 @@ def mcc_multiclass_loss(
     if cov_ypyp * cov_ytyt == 0:
         loss = torch.tensor(0.0)
```
Comment (davidegraff) [inline chemprop/train/loss_functions.py:149]: Should this line also be subtracted from 1?

**Review by davidegraff — COMMENTED** [review]
What about the case where the denominator is 0?

File: chemprop/train/loss_functions.py (hunk 1 above)
Comment (cjmcgill) [reply chemprop/train/loss_functions.py:149]: That loss is undefined, presumably due to bad coverage of the confusion matrix by the minibatch. I set it as a fixed zero so that one bad batch wouldn't break the model. It's already detached from the gradients so it won't matter for descent whether it's 1 or 0. But I suppose for training loss tracking it would be better for it to reflect a bad model. Will make that change along with fixing the testing values.

**Comment by davidegraff** [comment]
You would probably want the per-task mean of an MCC loss:
```python
if args.loss_function == "mcc":
    loss = loss.mean()
else:
    loss = loss.sum() / mask.sum()
```
This would be thematically similar to all other loss values which are the per-batch mean (more or less `loss.mean(-1).mean(-1)`)

File: chemprop/train/train.py [hunk 2]
```diff
@@ -50,65 +52,75 @@ def train(model: MoleculeModel,
...
+        if args.loss_function == "bounded_mse":
             lt_target_batch = lt_target_batch.to(torch_device)
             gt_target_batch = gt_target_batch.to(torch_device)
 
         # Calculate losses
-        if args.loss_function == 'mcc' and args.dataset_type == 'classification':
-            loss = loss_func(preds, targets, data_weights, mask) *target_weights.squeeze(0)
-        elif args.loss_function == 'mcc': # multiclass dataset type
+        if args.loss_function == "mcc" and args.dataset_type == "classification":
+            loss = loss_func(preds, targets, data_weights, mask) * target_weights.squeeze(0) * mask.sum()
```
Comment (davidegraff) [inline chemprop/train/train.py:91]: why multiply by `mask.sum()`? The MCC is already a batch-level value by its nature, so we shouldn't be modifying this value by the size of the batch. I think this might've gotten lost in my initial comment.

File: chemprop/train/train.py [hunk 3]
```diff
@@ -50,65 +52,75 @@ def train(model: MoleculeModel,
...
+        elif args.loss_function == "mcc":  # multiclass dataset type
             targets = targets.long()
             target_losses = []
             for target_index in range(preds.size(1)):
                 target_loss = loss_func(preds[:, target_index, :], targets[:, target_index], data_weights, mask[:, target_index]).unsqueeze(0)
                 target_losses.append(target_loss)
-            loss = torch.cat(target_losses).to(torch_device) * target_weights.squeeze(0)
-        elif args.dataset_type == 'multiclass':
+            loss = torch.cat(target_losses).to(torch_device) * target_weights.squeeze(0) * mask.sum()
```
Comment (davidegraff) [inline chemprop/train/train.py:98]:
```suggestion
            loss = torch.cat(target_losses) * target_weights.squeeze(0) * mask.sum()
```
same comment as above with the addition that the tensor should already be on the correct device

File: chemprop/train/loss_functions.py [hunk 4]
```diff
@@ -146,9 +146,10 @@ def mcc_multiclass_loss(
     cov_ytyt = n_samples**2 - torch.dot(t_sum, t_sum)
 
     if cov_ypyp * cov_ytyt == 0:
-        loss = torch.tensor(0.0)
+        loss = torch.tensor(1.0)
```
Comment (davidegraff) [inline chemprop/train/loss_functions.py:149]:
```suggestion
        loss = torch.tensor(1.0, device=torch_device)
```

File: chemprop/train/train.py (hunk 2 above)
Comment (cjmcgill) [reply chemprop/train/train.py:91]: Whoops sorry left this in by accident. Lost track of it when I updated the formatting.

**Review by davidegraff — APPROVED** [review]
lgtm
